docs/rfcs/0003-ecommerce-models.md

```diff
@@ -0,0 +1,136 @@
+If the user successfully completes a payment CyberSource will POST back to our order fulfillment
+API which will record the response in `Receipt` and change the `Order` status to `fulfilled`. If
+the status in the order POST back is not a success, `Order.status` will be `failed`.
+ 
+At some future time we will implement refund handling and this will set the `Order` status
+to `refunded`.
+
+#### Coupon
+
+A `Coupon` represents a coupon code and stores information about what kind of discount and
+to what product it can be applied. A `Coupon` may be used multiple times by multiple people,
+depending on the type of coupon.
+
+Once a `Coupon` is created it should **not** be edited. Unlike a `Product`, it is not expected
+that admins will edit fields of a `Coupon`. This is because a `Coupon` will be linked to from
+a `Basket` via `CouponBasket` and it provides information about the discount given to a user
+when they completed an order. `CouponBasket` and `CouponProduct` should also not be edited for
+the same reason.
+
+If an admin needs to update `Coupon` information it's recommended to set `enabled` to False on
+the existing coupons, then create new `Coupon`s to use instead.
+
+It has these fields:
+ - `coupon_code` - a unique string to be used by the end user
+ - `max_users` - Maximum number of users who can user the coupon
+ - `max_redemptions_per_user` - Maximum number of coupon redemptions per user. 
+ - `amount_type` - percent or fixed price discount
+ - `amount` - The amount of discount provided, depends on `amount_type`.
+ - `enabled` - If false, the coupon is not redeemable
+ - `expiration_date` - The latest time a coupon can be redeemed
+ - `activation_date` - The earliest time a coupon can be redeemed
+ 
+#### CouponProduct
+
+A `CouponProduct` is a link from a `Coupon` to `Product` describing which products the `Coupon`
+applies to. Unlike in MicroMasters a `Coupon` to a `Program` will **not** apply to `Course`
+or `CourseRun` which have the `Program` as a parent. Instead links to each of the items to be
+purchased must be explicitly created.
+
+#### CouponBasket
+
+A `CouponBasket` is a link from a `Coupon` to `Basket` describing the intent of a user to
+redeem the coupon on checkout (or that they already did). For right now there should be only
+one `CouponBasket` per `Basket` because only one coupon is allowed per purchase.
+
+### Audit tables
+
+#### OrderAudit
+
+The `OrderAudit` table will store append-only information about every change made to `Order`. It
+stores this information in a JSONField with information serialized from the linked `Basket`, each
+linked `Line`, and the `Coupon` used if any. There should be enough information to have a complete
+representation of the order for reporting purposes.
+
+Note that the `Order`, `Line` and `Coupon` tables should also keep accurate information
+suitable for reporting purposes. `OrderAudit` is meant to be a backup copy.
```

Comment:
We need a way to report on order transactions in some way, and our plan was to use the audit table. Does that make sense?

For example, in June George purchases a course. The June report includes a row for ann incoming payment for the course. In July, George asks for a refund. In the July report we need another row for the outgoing payment. I don't think the order table by itself is complete for reconciliation reports.